internal/controller/finalizer.go

```diff
@@ -0,0 +1,80 @@
+package controller
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type FinalizerHelper struct {
+	// runtimeClient is the client used to interact with the Kubernetes API
+	runtimeClient client.Client
+	// finalizer is the finalizer string to be added/removed
+	finalizer string
+}
+
+func NewFinalizerHelper(runtimeClient client.Client, finalizer string) FinalizerHelper {
+	fh := FinalizerHelper{
+		runtimeClient: runtimeClient,
+		finalizer:     finalizer,
+	}
+
+	return fh
+}
+
+func (fh FinalizerHelper) EnsureAdded(ctx context.Context, object client.Object) (bool, error) {
+	if controllerutil.ContainsFinalizer(object, fh.finalizer) {
+		// Finalizer already exists, no need to add it
+		return false, nil
+	}
+
+	logger := log.FromContext(ctx)
+
+	// We use a patch rather than an update to avoid conflicts when multiple controllers are adding their finalizer
+	// We use the patch from sigs.k8s.io/cluster-api/util/patch to handle the patching without conflicts
+	logger.Info("adding finalizer", "finalizer", fh.finalizer)
+	patchHelper, err := patch.NewHelper(object, fh.runtimeClient)
```

Comment:
I would move the patchHelper creation in the New function

Reply:
I think we should not use the same patchHelper for different objects, as its holding a copy of the given object to then calculate the difference between the new and old object.